File: _book/source/_posts/Algorithm/1.leetcode/tmp/46.全排列.py
#
# @lc app=leetcode.cn id=46 lang=python3
#
# [46] 全排列
#

# @lc code=start
from typing import List


class Solution:
    def permute2(self, nums: List[int]) -> List[List[int]]:
        import itertools
        return itertools.permutations(nums)

    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []

        def backtrack(nums, pre_list):
            if len(nums) <= 0:
                res.append(pre_list)
            else:
                for i in nums:
                    p_list = pre_list.copy()
                    p_list.append(i)
                    left_nums = nums.copy()
                    left_nums.remove(i)
                    backtrack(left_nums, p_list)

        backtrack(nums, [])
        return res

    def permute3(self, nums: List[int]) -> List[List[int]]:
        res = []
        length = len(nums)

        def backtrack(start=0):
            if start == length:
                res.append(nums[:])
            for i in range(start, length):
                #  1 3 2,  当第二次循环选择3，那么之后再改回2
                nums[start], nums[i] = nums[i], nums[start]
                backtrack(start + 1)
                nums[start], nums[i] = nums[i], nums[start]

        backtrack()
        return res


# @lc code=end

test = Solution()
print(test.permute3([1, 2, 3]))

File: source/_posts/Algorithm/经典算法/回溯/46.全排列.java
/*
 * @lc app=leetcode.cn id=46 lang=java
 *
 * [46] 全排列
 */

// @lc code=start
class Solution {

    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        // Arrays.sort(nums);
        backTrack(res, new ArrayList<>(), nums, new boolean[nums.length]);
        return res;
    }

    void backTrack(List<List<Integer>> res, List<Integer> tempList, int[] nums, boolean[] use) {
        if (tempList.size() == nums.length) {
            res.add(new ArrayList<>(tempList));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (use[i]) {
                continue;
            }
            use[i] = true;
            tempList.add(nums[i]);
            backTrack(res, tempList, nums, use);
            tempList.remove(tempList.size() - 1);
            use[i] = false;
        }
    }

    // public static void main(String[] args) {
    // Solution1 solution = new Solution1();
    // int[] candidates = new int[] { 1, 2, 2 };
    // solution.permuteUnique(candidates);
    // }
}
// @lc code=end
